### PR TITLE
[YUNIKORN-253] Add unit tests for the shim side API

### PR DESCRIPTION
**Pull request by adamantal**

### CODE REVIEW

**Comment by adamantal** [comment]
Ideally this should be merged in first before #146 so that I can add unit test coverage for the node types as well.

File: pkg/cache/context_test.go [hunk 1]
```diff
@@ -460,3 +464,60 @@ func TestRemoveTask(t *testing.T) {
 	// now there is no task left
 	assert.Equal(t, len(app.GetNewTasks()), 0)
 }
+
+func TestPublishEvents(t *testing.T) {
```
Comment (yangwwei) [inline pkg/cache/context_test.go:468]: we should have another UT to cover the case where app or task doesn't exist in the context.

File: pkg/cache/context_test.go [hunk 2]
```diff
@@ -460,3 +464,60 @@ func TestRemoveTask(t *testing.T) {
...
+			log.Logger.Info(event)
+			if strings.Contains(event, reason) && strings.Contains(event, message) {
+				found = true
+			}
+		default:
+			if !found {
+				t.Fatal("the event is supposed to be published")
+			}
+		}
+	}
```
Comment (yangwwei) [inline pkg/cache/context_test.go:522]: this could be hanging if something goes wrong. can you please add a timeout there?

Comment (adamantal) [reply pkg/cache/context_test.go:522]: You're right, created a constructor with`util.waitFor`.

File: pkg/cache/context_test.go (hunk 1 above)
Comment (adamantal) [reply pkg/cache/context_test.go:468]: right. added a negative test

File: pkg/cache/context_test.go [hunk 3]
```diff
@@ -20,20 +20,25 @@ package cache
 
 import (
 	"fmt"
+	"strings"
 	"testing"
 	"time"
 
+	"github.com/apache/incubator-yunikorn-k8shim/pkg/common/utils"
```
Comment (yangwwei) [inline pkg/cache/context_test.go:27]: pls format the import

File: pkg/cache/context_test.go [hunk 4]
```diff
@@ -460,3 +465,100 @@ func TestRemoveTask(t *testing.T) {
...
+	// check that the event has been published
+	select {
+	case event := <-recorder.Events:
+		log.Logger.Info(event)
+		if strings.Contains(event, reason) && strings.Contains(event, message) {
+			t.Fatal("event should not be published if the pod does not exist")
+		}
+	default:
+		break
+	}
```
Comment (yangwwei) [inline pkg/cache/context_test.go:506]: fix the potential issue that this may get stuck

**Comment by adamantal** [comment]
Rebased and fixed the review items - sorry for the delay. Could you please review it @yangwwei ?

**Comment by yangwwei** [comment]
hi @adamantal 

Changes look good. Can you fix the remaining lint error?
Next time you can run `make lint` locally before submitting the PR if you want to prevent such errors before creating the PR.

**Pull request merged**